test/integration/statefulset/util.go

```diff
@@ -0,0 +1,329 @@
+							},
+						},
+						{
+							Name: "home",
+							VolumeSource: v1.VolumeSource{
+								HostPath: &v1.HostPathVolumeSource{
+									Path: fmt.Sprintf("/tmp/%v", "home"),
+								},
+							},
+						},
+					},
+				},
+			},
+			ServiceName: "fake-service-name",
+			UpdateStrategy: v1beta1.StatefulSetUpdateStrategy{
+				Type: v1beta1.RollingUpdateStatefulSetStrategyType,
+			},
+			// Unnecessary to pass current tests
+			VolumeClaimTemplates: []v1.PersistentVolumeClaim{
+				// for volume mount "datadir"
+				newStatefulSetPVC("fake-pvc-name"),
+			},
+		},
+	}
+}
+
+func newStatefulSetPVC(name string) v1.PersistentVolumeClaim {
+	return v1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+			Annotations: map[string]string{
+				"volume.alpha.kubernetes.io/storage-class": "anything",
+			},
+		},
+		Spec: v1.PersistentVolumeClaimSpec{
+			AccessModes: []v1.PersistentVolumeAccessMode{
+				v1.ReadWriteOnce,
+			},
+			Resources: v1.ResourceRequirements{
+				Requests: v1.ResourceList{
+					v1.ResourceStorage: *resource.NewQuantity(1, resource.BinarySI),
+				},
+			},
+		},
+	}
+}
+
+// scSetup sets up necessities for Statefulset integration test, including master, apiserver, informers, and clientset
+func scSetup(t *testing.T) (*httptest.Server, framework.CloseFunc, *statefulset.StatefulSetController, informers.SharedInformerFactory, clientset.Interface) {
+	masterConfig := framework.NewIntegrationTestMasterConfig()
+	_, s, closeFn := framework.RunAMaster(masterConfig)
+
+	config := restclient.Config{Host: s.URL}
+	clientSet, err := clientset.NewForConfig(&config)
+	if err != nil {
+		t.Fatalf("error in create clientset: %v", err)
+	}
+	resyncPeriod := 12 * time.Hour
+	informers := informers.NewSharedInformerFactory(clientset.NewForConfigOrDie(restclient.AddUserAgent(&config, "statefulset-informers")), resyncPeriod)
+
+	sc := statefulset.NewStatefulSetController(
+		informers.Core().V1().Pods(),
+		informers.Apps().V1().StatefulSets(),
+		informers.Core().V1().PersistentVolumeClaims(),
+		informers.Apps().V1().ControllerRevisions(),
+		clientset.NewForConfigOrDie(restclient.AddUserAgent(&config, "statefulset-controller")),
+	)
+
+	return s, closeFn, sc, informers, clientSet
+}
+
+// Run STS controller and informers
+func runControllerAndInformers(sc *statefulset.StatefulSetController, informers informers.SharedInformerFactory) chan struct{} {
+	stopCh := make(chan struct{})
+	informers.Start(stopCh)
+	go sc.Run(5, stopCh)
+	return stopCh
+}
+
+func createHeadlessService(t *testing.T, clientSet clientset.Interface, headlessService *v1.Service) {
+	_, err := clientSet.Core().Services(headlessService.Namespace).Create(headlessService)
+	if err != nil {
+		t.Fatalf("Failed creating headless service: %v", err)
+	}
+}
+
+func createSTSsPods(t *testing.T, clientSet clientset.Interface, stss []*v1beta1.StatefulSet, pods []*v1.Pod) ([]*v1beta1.StatefulSet, []*v1.Pod) {
+	var ns string
+	if len(stss) != 0 {
```

Comment:
This is awkward. You can implement it based on this: https://github.com/kubernetes/kubernetes/blob/master/test/integration/replicaset/replicaset_test.go#L199-L218